### PR TITLE
run CWL 3step in centaur

### PR DESCRIPTION
**Pull request by danbills**
This adds 3step to the list of tests run by centaur.

Running this test exposed a regression that is fixed in this PR.

Unfortunately 1st-workflow is not suitable as it expects a file input but is added here anyway with the potential of running a modified version.

### CODE REVIEW

File: centaur/src/main/resources/standardTestCases/three_step/three_step.cwl [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    - valueFrom: $(inputs.pattern)
+      shellQuote: false
+    - valueFrom: ${return inputs.file}
+      shellQuote: false
+    - valueFrom: '|'
+      shellQuote: false
+    - valueFrom: wc
+      shellQuote: false
+    - valueFrom: -l
+      shellQuote: false
```
Comment (mcovarr) [inline centaur/src/main/resources/standardTestCases/three_step/three_step.cwl]: is there really not a more concise expression for this in CWL? or is this the more common form that we want to focus on supporting?

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 2]
```diff
@@ -61,15 +61,7 @@ case class InputParameterCommandPart(commandInputParameter: CommandInputParamete
...
-        Case where the only input binding option specified is the position.
-
-        NB: We ignore the position as these have already been sorted prior to being submitted as command part
-
-        In the cases where we have a defined inputbinding with more options,
-        we should consider instantiating a CommandLineBindingCommandPart and delegating this call.
-      */
-      case CommandInputParameter(commandInputParamterFqn, _,_,_,_,_, Some(CommandLineBinding(None,Some(_),None,None,None,None,None)),_,_) =>
+      case CommandInputParameter(commandInputParamterFqn, _,_,_,_,_, _,_,_) =>
         inputsMap.get(LocalName(FullyQualifiedName(commandInputParamterFqn).id)) match {
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: or at least correct the preexisting misspelling of `parameter` 😛

File: cwl/src/test/scala/cwl/CommandLineToolSpec.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  it should "order input arguments when bound to command line" in {
+
+    val a = CommandInputParameter("a", inputBinding = None)
+    val b = CommandInputParameter("b", inputBinding = Some(CommandLineBinding()))
+
+    val ordered: Seq[CommandInputParameter] = CommandLineTool.orderedForCommandLine(Array(a,b))
+
+    ordered.contains(a) shouldBe false
+
+    ordered.contains(b) shouldBe true
```
Comment (mcovarr) [inline cwl/src/test/scala/cwl/CommandLineToolSpec.scala]: was this actually a test of ordering?

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 4]
```diff
@@ -120,18 +120,20 @@ case class CommandLineTool private(
...
+object CommandLineTool {
+
   /**
     * Sort according to position. If position does not exist, use 0 per spec:
     * http://www.commonwl.org/v1.0/CommandLineTool.html#CommandLineBinding
     */
-  def orderedInputs: Seq[CommandPart] =
+  protected[cwl] def orderedForCommandLine(inputs: Array[CommandInputParameter]):Seq[CommandInputParameter] = {
     inputs.
+      filter(_.inputBinding.isDefined).
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/CommandLineTool.scala]: expanding on the scaladoc above, something like "and if there is no input binding ignore the input parameter"

File: centaur/src/main/resources/standardTestCases/first_workflow/arguments.cwl [hunk 5]
```diff
@@ -0,0 +1,22 @@
+cwlVersion: "v1.0"
+class: "CommandLineTool"
+label: "Example trivial wrapper for Java 7 compiler"
+hints: 
+  - dockerPull: "java:7-jdk"
```
[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/first_workflow/arguments.cwl]: The latter

File: centaur/src/main/resources/standardTestCases/three_step/three_step.cwl (hunk 1 above)
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/three_step/three_step.cwl]: Afaik this is it but I have minimal experience in CWL writing

File: cwl/src/test/scala/cwl/CommandLineToolSpec.scala (hunk 3 above)
Comment (danbills) [reply cwl/src/test/scala/cwl/CommandLineToolSpec.scala]: No of filtering.  I'll fix it up

File: engine/src/main/resources/swagger/cromwell.yaml [hunk 6]
```diff
@@ -586,8 +586,6 @@ parameters:
     type: string
     description: The workflow language for the file you submitted. Cromwell currently supports WDL.
     required: false
-    enum: [wdl, cwl]
-    default: wdl
```
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/resources/swagger/cromwell.yaml]: Unrelated but very desirable

File: cwl/src/test/scala/cwl/CommandLineToolSpec.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import cwl.CommandLineTool.CommandInputParameter
+import org.scalatest.{FlatSpec, Matchers}
+
+class CommandLineToolSpec extends FlatSpec with Matchers {
+
+  behavior of "CommandLineTool"
+
+  it should "order input arguments when bound to command line" in {
+
+    val a = CommandInputParameter("a", inputBinding = None)
```
Comment (cjllanwarne) [inline cwl/src/test/scala/cwl/CommandLineToolSpec.scala]: OOI, what is this case where `inputBinding` is `None`?

File: engine/src/main/resources/swagger/cromwell.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply engine/src/main/resources/swagger/cromwell.yaml]: Wait, is this no longer true?

[the remark this answers is not in the document]
Comment (cjllanwarne) [reply engine/src/main/resources/swagger/cromwell.yaml]: Side note: this swagger probably wants `workflowTypeVersion` too?

File: centaur/src/main/resources/standardTestCases/first_workflow.test [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# This test is ignored as it expects a local file to exist, and we avoid that in centaur to allow cloud/local runs to just work.
+# In order to facilitate running this test, a "make File" step could be added and the "file" input could be removed.
```
[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/first_workflow.test]: Ok

File: engine/src/main/resources/swagger/cromwell.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (danbills) [reply engine/src/main/resources/swagger/cromwell.yaml]: @cjllanwarne it doesn't work!  It always submits the default 😢

File: centaur/src/main/resources/standardTestCases/first_workflow.test [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+name: firstWorkflowCwl
+testFormat: workflowsuccess
+workflowType: CWL
+workflowTypeVersion: v1.0
+
+files {
+  wdl: first_workflow/1st-workflow.cwl
+  inputs: first_workflow/inputs.json
+  imports: [
+    first_workflow/arguments.cwl,
```
[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/first_workflow.test]: theirs

File: cwl/src/test/scala/cwl/CommandLineToolSpec.scala (hunk 7 above)
Comment (danbills) [reply cwl/src/test/scala/cwl/CommandLineToolSpec.scala]: An element could be used in an expression and not included in the command line script.

File: centaur/src/main/resources/standardTestCases/first_workflow.test [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# This test is ignored as it expects a local file to exist, and we avoid that in centaur to allow cloud/local runs to just work.
+# In order to facilitate running this test, a "make File" step could be added and the "file" input could be removed.
+ignore: true
+name: firstWorkflowCwl
+testFormat: workflowsuccess
+workflowType: CWL
+workflowTypeVersion: v1.0
+
+files {
+  wdl: first_workflow/1st-workflow.cwl
```
[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/first_workflow.test]: indeed.  I think that's another PR of much noise though.

File: centaur/src/main/resources/standardTestCases/first_workflow.test [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+  inputs: first_workflow/inputs.json
+  imports: [
+    first_workflow/arguments.cwl,
+    first_workflow/tar-param.cwl
+  ]
+}
+
+metadata { }
```
[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/first_workflow.test]: I will add them as expected but comment them out as we do not yet support CWL-specific outputs

File: centaur/src/main/resources/standardTestCases/three_step.test [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+testFormat: workflowsuccess
+workflowType: CWL
+workflowTypeVersion: v1.0
+
+files {
+  wdl: three_step/three_step.cwl
+  inputs: three_step/inputs.json
+}
+
+metadata { }
```
Comment (cjllanwarne) [inline centaur/src/main/resources/standardTestCases/three_step.test]: Expectations here too?

File: cwl/src/test/scala/cwl/CommandLineToolSpec.scala (hunk 3 above)
Comment (cjllanwarne) [reply cwl/src/test/scala/cwl/CommandLineToolSpec.scala]: Were you going to add an ordering test too?

**Comment by cjllanwarne** [comment]
@danbills there's another centaur test `workflow_type_and_version_cwl` - is that related to- or maybe an older version of - this one?

File: centaur/src/main/resources/standardTestCases/three_step.test (hunk 12 above)
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/three_step.test]: As I understand your PR will facilitate the emitting of outputs.  When merged I will add meaningful expectations.

**Comment by danbills** [comment]
@cjllanwarne I see the `workflow_type_and_version_cwl` test but so far as I can tell it is testing that such parameters are accepted by cromwell.  

It is curious that it passes though because it is actually passing in a WDL file!  We should likely disallow this.

**Comment by danbills** [comment]
I merged this as it fixes a regression in the code.

However, there is an issue w/ centaur in running a test on _only_ centaurLocal.

I tried:
* `backends: [Local] which runs on both TES and JES and fails on those.
* `backends: [Local] and `tags: [localdockertest]` which runs on both TES and JES and fails on those.
* `backends: [Local]`, `tags: [localdockertest]` and `backendsMode: "only"` which ignores the test in all environments.

I merged in this final state as it at least does not fail the build overall.

**Pull request merged**